### ISSUE TITLE
Allow full replacement of default marker icons

### ISSUE THREAD

**Issue by denydias** [Leaflet/Leaflet#4604]
This is a feature request, not a bug.

As per #3696, Leaflet's `L.Icon.Default` only allows changing `iconUrl` and `shadowUrl`. It would be VERY nice if Leaflet also support full `L.Icon.Default` replacement. This would allow, for instance:

``` js
L.Icon.Default = L.Icon.Default.extend({
  options: {
    iconUrl: 'my-new-marker-icon.png',
    iconRetinaUrl: 'my-new-marker-icon@2x.png',
    shadowUrl: 'my-new-marker-shadow.png',
    shadowRetinaUrl: 'my-new-marker-shadow@2x.png',
    iconSize: [ 35, 45 ],
    iconAnchor: [ 17, 42 ],
    popupAnchor: [ 1, -32 ],
    shadowAnchor: [ 10, 12 ],
    shadowSize: [ 36, 16 ],
    className: "my-new-css-class"
  }
});
```

Or even better, although requiring deeper changes:

``` js
L.Icon.Default = L.ExtraMarkers.icon({
  icon: 'fa-home',
  markerColor: 'blue',
});
```

`L.ExtraMarkers.icon()` comes from [Leaflet.extra-markers plugin](https://github.com/coryasilva/Leaflet.ExtraMarkers) (but it could be [Leaflet.awesome-markers](https://github.com/lvoogdt/Leaflet.awesome-markers) or any other handling marker icons).

This is mostly desired for projects sourcing Leaflet from `https://cdnjs.cloudflare.com`, as there is no way to change the source code.

**Comment by denydias** [comment]
It's so nice to see how this issue evolved! Nice work, @IvanSanchez et all.

**Comment by ghybs** [comment]
Hi,

It seems to me that it is actually quite easy to change the default icon already, but there _is_ a trap in that `L.Icon.Default` is referenced in `L.Marker` options at script initialization.
Therefore you cannot just re-assign `L.Icon.Default`  (the `L.Marker` option will keep the original reference)

You just have to override that `L.Marker` option with your desired icon:
```javascript
L.Marker.mergeOptions({
  icon: L.ExtraMarkers.icon({ // whatever you like
    icon: 'glyphicon glyphicon-heart',
    markerColor: 'blue',
  })
});
```

Demo: http://playground-leaflet.rhcloud.com/boda/1/edit?html,output

If many people are looking for such replacement, it might be worth mentioning it in the [docs](http://leafletjs.com/reference-1.0.0.html#icon-default-option).

**Comment by denydias** [comment]
Quite interesting, @ghybs. Your tip works with Leaflet 0.7.7 too. Cool!

I agree that this might be a good addition to the docs.

**Comment by johnd0e** [comment]
Close, as resolved a long ago by #4605
